lib/hash.h

```diff
@@ -11,6 +11,8 @@
 #include <tuple>
 #include <type_traits>
 
+#include "lib/big_int_util.h"
```

Comment:
maybe it would make sense to reverse the dependency and instantiate hasher inside `lib/big_int_util.h`? Right now it seems that everyone who'd like to use `lib/hash.h` would pull `big_int_util.h`, this seems to be not quite a good thing to me.

Reply:
`big_int_util` is mostly used for the declaration of `big_int`, which is used fairly pervasively in the compiler. We can split that out?

Reply:
Right. But why does it become the hash dependency? It seems it should be the opposite: `big_int_utils` should depend on `hash` and not vice versa.

Reply:
Yes, I have a fix for that in #4655.